src/recon_buildblock/BinNormalisationPETFromComponents.cxx

```diff
@@ -173,6 +173,20 @@ BinNormalisationPETFromComponents::allocate(
 
   this->_already_allocated = true;
 }
+void
+BinNormalisationPETFromComponents::allocate(shared_ptr<const ProjDataInfo> pdi_sptr, MLEstimateComponentBasedNormalisation& normalization_estimator)
+{
+  if (!normalization_estimator.has_processed_data())
+    {
+      error("BinNormalisationPETFromComponents: internal error: allocate called on MLEstimateComponentBasedNormalisation "
+            "without it having processed the data");
+    }
+  this->proj_data_info_sptr = pdi_sptr;
+  this->efficiencies = *normalization_estimator.get_efficiencies();
+  this->geo_data = *normalization_estimator.get_geo_data();
+  this->block_data = *normalization_estimator.get_block_data();
+  this->_already_allocated = true;
+}
```

Comment:
I believe this is the cleanest way to implement all these changes. Here is a use case from my cpp program.

``` cpp
    MLEstimateComponentBasedNormalisation ml_estimator(out_filename_tmp, *measured_projdata, *model_activity_projdata,
                                                       num_eff_iterations, num_iterations, do_geo, do_block,
                                                       do_symmetry_per_block, do_KL, do_display, do_save_to_file);
    ml_estimator.process();

    // Apply the normalization factors
    BinNormalisationPETFromComponents bin_normalisation;
    bin_normalisation.allocate(measured_projdata->get_proj_data_info_sptr(), ml_estimator);
    bin_normalisation.set_up(measured_projdata->get_exam_info_sptr(), measured_projdata->get_proj_data_info_sptr());

    // Compute the projdata
    ProjDataInMemory normalization_projdata(*measured_projdata);
    normalization_projdata.fill(1.F);
    bin_normalisation.undo(normalization_projdata);
```

Reply:
I acknowledge https://github.com/UCL/STIR/issues/1498#issuecomment-2322294430, however this does work and basically provides an alternative setter of the fields.

Reply:
I'd rather do it the other way around I'm afraid. Your way is very different from anything else in STIR.

I very much prefer to let `MLEstimateComponentBasedNormalisation` have a `shared_ptr<BinNormalisationPETFromComponents> normalisation_sptr` member, which we can get via `get_output()` (returning a `shared_ptr<const BinNorm...>`). The `MLEstimateComponentBasedNormalisation::get_efficiencies()` would then just forward to `normalisation_sptr->get_efficiencies()` (note that I wouldn't let those return `shared_ptr` then, but references). Any reason not to do it this way?